File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:

    void helper(vector<int>& candidates,
                vector<vector<int>>& ans,
                vector<int>& temp,
                int start,
                int target)
    {
        if (target == 0)
        {
            ans.push_back(temp);
            return;
        }

        for (int i = start; i < candidates.size(); i++)
        {
            if (candidates[i] > target)
                continue;

            temp.push_back(candidates[i]);

            helper(candidates,
                   ans,
                   temp,
                   i,
                   target - candidates[i]);

            temp.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target)
    {
        vector<vector<int>> ans;
        vector<int> temp;

        helper(candidates, ans, temp, 0, target);

        return ans;
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:

        candidates.sort()

        res = []
        temp = []

        def helper(candidates, res, temp, remain, start):

            # Target reached
            if remain == 0:
                res.append(temp.copy())
                return

            # Try all candidates
            for i in range(start, len(candidates)):

                # Skip duplicate choices at the same level
                if i > start and candidates[i] == candidates[i - 1]:
                    continue

                # Since sorted, no later number can work
                if candidates[i] > remain:
                    break

                # Choose
                temp.append(candidates[i])

                # Explore
                # i + 1 because each element can be used only once
                helper(candidates, res, temp, remain - candidates[i], i + 1)

                # Backtrack
                temp.pop()

        helper(candidates, res, temp, target, 0)

        return res